### DISCUSSION TITLE
WithPagination doesn't work properly with dependency Injection on Mount()

### DISCUSSION THREAD

**Discussion by Allanfd12** [livewire/livewire#4974] [1. Help]
I've tried to inject the dependency of IProductRepository on the ShopComponent and use it in the render function, but WithPagination calls render function directly without inject properly the dependencys on mount(), instead, when I tried to inject the dependency by __construct, it works on WithPagination request, but does't work on firt load request

Any suggestion on how to solve it, or if this is a problem of the livewire ?

the code:
![image](https://user-images.githubusercontent.com/75325265/169176390-bb4d6af2-a906-4725-97ac-bb01ff792dae.png)
the error: 
![image](https://user-images.githubusercontent.com/75325265/169176922-29e9f169-e900-4921-89ee-f6f8cf9be19f.png)

the full project: 
https://github.com/Allanfd12/ecommerce-app

**Comment by joaopalopes24** [comment]
Try putting the `private IProductRepository $productRepository;` attribute as public and test again.

**Reply by Allanfd12** [reply]
![image](https://user-images.githubusercontent.com/75325265/169914922-1c8cebaf-4a17-46d7-bd48-f8474032ef57.png)
change to public don't change de error, actually, the problem seems to be caused by the livewire calling the render() method before the mont()

and on first load, the livewire says that the property needs to be private
![image](https://user-images.githubusercontent.com/75325265/169915585-214733a4-aea7-4dc2-a35b-16cfc552ca3d.png)

the solution i found was to call de dependency in render() method, like this:
![image](https://user-images.githubusercontent.com/75325265/169921772-683907eb-4982-49b0-95f4-c831cae4f700.png)

**Comment by joaopalopes24** [comment]
Another point, I recommend you use computed properties instead of passing the attributes `(products and categories)` the way you are doing.

**Reply by Allanfd12** [reply]
do you have an example of how to do this? the example from the livewire documentation is not so enlightening.

**Comment by joaopalopes24** [comment]
Would be like this:

```
use WithPagination;

    public int $perPage = 10;

    public string $sortMethod =  'default';

    public $formater = FormaterService::class;

    private IProductRepository $productRepository;

    public function getCategoriesProperty(): Collection
    {
        return Category::all();
    }

    public function getProductsProperty(): LengthAwarePaginator
    {
        return IProductRepository::query()->paginate($this->perPage);
    }

    public function store(int $product_id, int $quantity = 1): RedirectResponse
    {
        CartService::add(Product::find($product_id),$quantity);

        ViewNotifierService::success(DefaultMessagesSuccess::CartProductAdd);

        return redirect()->route('cart');
    }

    public function render()
    {
        return view('livewire.shop-component')->layout('layouts.padrao');
    }
```

I do not recommend using Repositories.

**Reply by joaopalopes24** [reply]
I recommend typing the input and output attributes, and organizing the code like `Traits` are always imported at the beginning, and the other methods are always between `mount` and `render`.
